geth/node/peerpool.go

```diff
@@ -0,0 +1,291 @@
+/*
+Peer pool works in a following way:
+1. For each configured topic it will create discovery v5 search query.
+Search query will periodically do regular kademlia lookups with bucket size 16
+and send a topic query to every node that is returned from kademlia lookup.
+Eventually nodes with required topics will be found and we will pass them to p2p server.
+2. Additional loop will be created for every topic that will synchronize
+found nodes with p2p server. This loop will follow next logic:
+- if node is found and max limit of peers is not reached we will add this node to
+  server and assume that it is connected
+- if max limit is reached we will add peer to our peer topic table for later use
+- if min limit is reached - frequency will be changed to a keepalive timer, this is required cause we need
+  frequent lookups only when we are looking for a peer
+3. when peer is disconnected we do 3 things:
+  - select new peer from peers table that was updated no longer than foundTimeout and not
+    in the connected (90s)
+  - set a peer as not connected
+  - check how many peers do we have and in case if we went below min limit - set period to fastSync
+*/
+
+package node
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/discv5"
+	"github.com/status-im/status-go/geth/db"
+	"github.com/status-im/status-go/geth/params"
+)
+
+var (
+	// ErrDiscv5NotRunning returned when pool is started but discover v5 is not running or not enabled.
+	ErrDiscv5NotRunning = errors.New("Discovery v5 is not running")
+)
+
+const (
+	foundTimeout    = 60 * time.Minute
+	defaultFastSync = 500 * time.Millisecond
+	defaultSlowSync = 30 * time.Minute
+)
+
+// NewPeerPool creates instance of PeerPool
+func NewPeerPool(config map[discv5.Topic]params.Limits, fastSync, slowSync time.Duration, cache *db.PeersDatabase) *PeerPool {
+	return &PeerPool{
+		config:   config,
+		fastSync: fastSync,
+		slowSync: slowSync,
+		cache:    cache,
+	}
+}
+
+type peerInfo struct {
+	// discoveredTime last time when node was found by v5
+	discoveredTime mclock.AbsTime
+	// connected is true if node is added as a statis peer
+	connected bool
+	// dropped is true if node was dropped by p2p server
+	// it can become false only when kademlia will discover it again
+	dropped bool
```

Comment:
can it be both connected and dropped at the same time?

Reply:
it can be:
connected=false  dropped=false // peer added when max limit is already reached
connected=true  dropped=false // peer added when min limit is not reached
connected=false  dropped=true // peer was dropped by p2p server

Reply:
what about `connected=true dropped=true`?

Reply:
it becomes connected=false if it was dropped, i think i can remove it from table instead of adding dropped variable

Reply:
Yeah, that sounds good!
I'm very cautious about using bools to implement state machines. You always end up with 2^n states, and that is a huge source of potential bugs.